x-pack/solutions/observability/plugins/streams/server/lib/streams/client.ts

```diff
@@ -0,0 +1,749 @@
+
+    if (isEnabled) {
+      return { acknowledged: true, result: 'noop' };
+    }
+
+    await this.upsertStream({
+      definition: rootStreamDefinition,
+    });
+
+    return { acknowledged: true, result: 'created' };
+  }
+
+  /**
+   * Disabling streams means deleting the logs root stream
+   * AND its descendants, including any Elasticsearch objects,
+   * such as data streams. That means it deletes all data
+   * belonging to wired streams.
+   *
+   * It does NOT delete ingest streams.
+   */
+  async disableStreams(): Promise<DisableStreamsResponse> {
+    const isEnabled = await this.isStreamsEnabled();
+    if (!isEnabled) {
+      return { acknowledged: true, result: 'noop' };
+    }
+
+    const definition = await this.getStream(LOGS_ROOT_STREAM_NAME);
+
+    await this.deleteStreamFromDefinition(definition);
+
+    return { acknowledged: true, result: 'deleted' };
+  }
+
+  /**
+   * Resyncing streams means re-installing all Elasticsearch
+   * objects (index and component templates, pipelines, and
+   * assets), using the stream definitions as the source of
+   * truth.
+   *
+   * Streams are re-synced in a specific order:
+   * the leaf nodes are synced first, then its parents, etc.
+   * Thiis prevents us from routing to data streams that do
+   * not exist yet.
+   */
+  async resyncStreams(): Promise<ResyncStreamsResponse> {
+    const streams = await this.getManagedStreams();
+
+    const streamsWithDepth = streams.map((stream) => ({
+      stream,
+      depth: stream.name.match(/\./g)?.length ?? 0,
+    }));
+
+    const streamsInOrder = orderBy(streamsWithDepth, 'depth', 'desc');
+
+    for (const { stream } of streamsInOrder) {
+      await this.syncStreamObjects({
+        definition: stream,
+      });
+    }
+    return { acknowledged: true, result: 'updated' };
+  }
+
+  private async syncStreamObjects({ definition }: { definition: StreamDefinition }) {
+    const { assetClient, logger, scopedClusterClient } = this.dependencies;
+
+    if (isWiredStream(definition)) {
+      await syncWiredStreamDefinitionObjects({
+        definition,
+        logger,
+        scopedClusterClient,
+      });
+    } else if (isIngestStream(definition)) {
+      await syncIngestStreamDefinitionObjects({
+        definition,
+        scopedClusterClient,
+        logger,
+        dataStream: await this.getDataStream(definition.name),
+      });
+    }
+
+    await assetClient.syncAssetList({
+      entityId: definition.name,
+      entityType: 'stream',
+      assetType: 'dashboard',
+      assetIds: definition.dashboards ?? [],
+    });
+  }
+
+  /**
+   * Creates or updates a stream. The routing of the parent is
+   * also updated (including syncing to Elasticsearch).
+   */
+  async upsertStream({
+    definition,
+  }: {
+    definition: StreamDefinition;
+  }): Promise<UpsertStreamResponse> {
+    const { result, parentDefinition } = await this.validateAndUpsertStream({
+      definition,
+    });
+
+    if (parentDefinition) {
```

Comment:
Something weird is going on here - I didn't check what's going on, but this branch is required so I can create a new non-existing child via `PUT` and it will be added to the parents routing array, right?

This is not happening though - if I
```
PUT kbn:/api/streams/logs.atest
{
  "ingest": {
    "processing": [],
    "routing": [],
    "wired": {
      "fields": {}
    }
  }
```

then on `GET kbn:/api/streams` , it won't be part of the `routing` array.

I think something is off here.

Reply:
do you mean `GET kbn:/api/streams/logs`?

Reply:
Yes, I do, sorry

Reply:
Taking the same steps as you, I first had an issue with enabling logs when the asset index exists, but the streams one doesn't. After fixing that one, the next bug I ran into was an invalid painless script being generated when the condition is undefined - I'm not sure how that bug crept up in here, looking into it. After fixing that one, it does work.

Reply:
ah, maybe this was never working before? It looks like it would have failed with a 404 when updating the parent stream

Reply:
possible it broke a while ago - definitely a good edge case for an integration test, thanks for adding it to the other issue

Reply:
Are you trying to do the `PUT` before it's been forked? What would the criteria be for the `log` reroute?

Reply:
the criteria is empty in this case, which means no data is routed to the stream, but it's still there. I think this functionality will be really handy down the line, because it allows the user to "close the tap" on data flowing a certain way without deleting the whole structure.